app.py

```diff
@@ -0,0 +1,423 @@
+            #   [ -p <processing options> (default = none)
+            #
+            input_cmd = "%s -n \"%s@CLIP\" -h %s -W %s -H %s -L" % (self.engine.get_read_frame_path(),
+                                                                    full_path,
+                                                                    "%s:Gateway" % self.engine.get_server_hostname(), 
+                                                                    info["width"],
+                                                                    info["height"])
+            
+            thumbnail_jpg = os.path.join(self.engine.get_backburner_tmp(), "tk_thumb_%s.jpg" % uuid.uuid4().hex)
+            full_cmd = "%s > %s" % (input_cmd, thumbnail_jpg)
+            self.log_debug("Executing %s" % full_cmd)
+            if os.system(full_cmd) != 0:
+                self.log_warning("Could not extract thumbnail! See error log for details.")
+            else:
+                self.log_debug("Wrote thumbnail %s" % thumbnail_jpg)
+                self.sgtk.shotgun.upload_thumbnail(sg_data["type"], sg_data["id"], thumbnail_jpg)
+                self.log_debug("Uploaded thumbnail to shotgun.")
+                try:
+                    os.remove(thumbnail_jpg)
+                    self.log_debug("Removed temporary file '%s'." % thumbnail_jpg)
+                except Exception, e:
+                    self.log_warning("Could not remove temporary file '%s': %s" % (thumbnail_jpg, e))    
+                
+            # thumbnail upload done!
+
+        # now start the version creation process
+        self.log_debug("Will associate upload with shotgun entity %s..." % sg_data)
+
+        # create a version in Shotgun
+        if info["versionNumber"] != 0:
+            title = "%s v%03d" % (info["sequenceName"], info["versionNumber"])
+        else:
+            title = info["sequenceName"]
+        
+        data = {}
+        data["code"] = title
+        data["description"] = comments
+        data["project"] = self.context.project
+        data["entity"] = sg_data
+        data["created_by"] = self.context.user
+        data["user"] = self.context.user
+        
+        # general metadata for the version
+        # for the frame range, there isn't very meaningful metadata we can add
+        # and we don't have corresponding frames on disk
+        # so set the first frame to 1 in order to normalize the frames from flame
+        # which typically start at 10:00:00.00
+        #
+        # also note that flame is out-exclusive, meaning that if you have the 
+        # frame range 100-111, it corresponds to the frames
+        # 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110
+        # 
+        # We transform the above frame range (100-111) to be 1-10 in shotgun with length 10. 
+        #
+        data["sg_first_frame"] = 1
+        data["sg_last_frame"] = info["sourceOut"] - info["sourceIn"]
+        data["frame_count"] = info["sourceOut"] - info["sourceIn"]
```

Comment:
Pretty minor : it might be better to use sg_last_frame and sg_first_frame to compute frame_count. So, if the method to compute those ones change, this line does not have to be changed.